cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java

```diff
@@ -514,15 +515,17 @@ private void createApplicationJar(ApplicationBundler bundler,
     }
 
     // The location name is computed from the MD5 of all the classes names
-    // The localized name is always APPLICATION_JAR
+    // The localized name is always APPLICATION_JAR when the job is launched using SSH.
     List<String> classList = classes.stream().map(Class::getName).sorted()
         .collect(Collectors.toList());
     Hasher hasher = Hashing.md5().newHasher();
     for (String name : classList) {
       hasher.putString(name);
     }
-    // Only depends on class list so that it can be reused across different launches
-    String name = hasher.hash().toString() + "-" + Constants.Files.APPLICATION_JAR;
+    hasher.putString(ProjectInfo.getVersion().toString());
+    // Only depends on class list and cdap version so that it can be reused across different launches
+    String hashVal = hasher.hash().toString();
+    String name = hashVal + "-" + Constants.Files.APPLICATION_JAR;
```

Comment:
can we change format to application_hash.jar to be consistent with other hashed file names.

Reply:
use String.format

Reply:
yeah this is how we are localizing it to cache it in GCS: https://github.com/cdapio/cdap/pull/14988/files#diff-47d216c5d83d7f34c4046bc4801f7b0be5e2521fe14115e05b052cc114ca03cfR127

Locally it is cached as `hashVal + "-" + Constants.Files.APPLICATION_JAR`.

Reply:
Do we also want to change the name locally?